source/s3_meta_request.c

```diff
@@ -2155,6 +2169,160 @@ void aws_s3_meta_request_result_setup(
+                eof,
+                previously_buffered);
+
+            meta_request->async_write.eof = eof;
+            meta_request->async_write.synced_future = aws_future_void_acquire(write_future);
+            ready_to_send = true;
+        } else {
+            /* Can't send yet. Complete write's future, so we can get more data */
+            AWS_LOGF_TRACE(
+                AWS_LS_S3_META_REQUEST,
+                "id=%p: write(data=%zu, eof=%d) previously-buffered=%zu. Not enough to upload.",
+                (void *)meta_request,
+                data.len,
+                eof,
+                previously_buffered);
+
+            /* TODO: does a future that completes immediately risk stack overflow?
+             * If a user does tiny writes, and registers callbacks on the write-future,
+             * they'll fire synchronously. If the user repeats, the stack will just grow and grow. */
+            aws_future_void_set_result(write_future);
+        }
+    }
+
+    if (illegal_usage_terminate_meta_request) {
+        aws_future_void_set_error(write_future, AWS_ERROR_INVALID_STATE);
+        aws_s3_meta_request_set_fail_synced(meta_request, NULL, AWS_ERROR_INVALID_STATE);
+    }
+
+    aws_s3_meta_request_unlock_synced_data(meta_request);
+    /* END CRITICAL SECTION */
+
+    if (ready_to_send || illegal_usage_terminate_meta_request) {
+        /* Schedule the work task, to continue processing the meta-request */
+        aws_s3_client_schedule_process_work(meta_request->client);
+    }
+
+    return write_future;
+}
+
+/* Copy pending async-write data into the buffer.
+ * This is only called when there's enough data for the next part. */
+static int s_s3_meta_request_read_from_pending_async_writes(
+    struct aws_s3_meta_request *meta_request,
+    struct aws_byte_buf *dest,
+    bool *eof) {
+
+    *eof = false;
+
+    /* This read should NOT happen unless there's a pending async-write */
+    AWS_ASSERT(meta_request->async_write.synced_future != NULL);
+
+    aws_byte_buf_write_to_capacity(dest, &meta_request->async_write.data_cursor);
+
+    /* We should have filled the dest buffer, unless this is the final write */
+    AWS_ASSERT(dest->len == dest->capacity || meta_request->async_write.eof);
+
+    /* If there's still enough data to fill another part, return success.
+     * We'll copy more the next time this function is called. */
+    if (meta_request->async_write.data_cursor.len >= meta_request->part_size) {
+        return AWS_OP_SUCCESS;
+    }
+
+    /* Otherwise, we're done with the current async-write... */
+
+    if (meta_request->async_write.data_cursor.len == 0) {
+        /* We've sent all the data */
+        meta_request->async_write.data_buffer.len = 0;
+
+        if (meta_request->async_write.eof) {
+            *eof = true;
+        }
+    } else {
+        /* There's leftover data. Copy it to the front of the buffer.
+         * We'll send it after we get more writes. */
+        memcpy(
```

Comment:
concern: potential ub with overlapping src and dest?

Reply:
We only get here if we had 1+ parts worth of data, and we're only doing the memcpy if there's `< 1` part's worth of data leftover, should we *shouldn't* have a problem.

But I didn't know about that undefined behavior. I"ll add some asserts